### PR TITLE
tables: disable UpdateDeltaForTable if TxnCtx is nil

### PR DESCRIPTION
**Pull request by kennytm**
### What problem does this PR solve? 

This is a follow-up to pingcap/tidb-lightning#274. 

In the flamegraph of `AddRecord()` (with a special session used by Lightning), it was found that calculation used for `(*TransactionContext).UpdateDeltaForTable()` occupy 27% run time (!) of the entire `AddRecord()` call. The delta info is useless for Lightning, so we need a way to disable it.

<details><summary>Flamegraph for Lightning before this PR</summary>

Annotation 2020-03-01 033907-fs8

</details>

### What is changed and how it works?

Lightning will set `sessVar.TxnCtx = nil` as a signal that `UpdateDeltaForTable` is not needed. This PR adds the `nil` check in `AddRecord()` for skipping the computation. (`UpdateRecord()` and `RemoveRecord()` are not used by Lightning and thus their use of `UpdateDeltaForTable` are untouched.)

<details><summary>Flamegraph for Lightning after this PR</summary>

Annotation 2020-03-01 034115-fs8

</details>

The time needed to encode one row of TPC-C "CUSTOMER" table (21 columns, 1 index) went down from 10.5&nbsp;ms/op to 8.2&nbsp;ms/op.

Note that this PR does *not* affect TiDB itself, since `sessVar.TxnCtx` should always be non-nil in normal use. Thus no test cases were added in TiDB itself. (Test for correctness will be in Lightning.)

### Check List 

Tests 

 - Unit test

Code changes

Side effects

Related changes

Release note

### CODE REVIEW

**Comment by kennytm** [comment]
PTAL @bobotu cc @3pointer @siddontang

**Comment by GregoryIan** [comment]
Cool! LGTM

**Comment by GregoryIan** [comment]
@bobotu @lysu PTAL

File: table/tables/tables.go [hunk 1]
```diff
@@ -590,15 +590,19 @@ func (t *TableCommon) AddRecord(ctx sessionctx.Context, r []types.Datum, opts ..
 		}
 	}
 	sc.AddAffectedRows(1)
-	colSize := make(map[int64]int64, len(r))
-	for id, col := range t.Cols() {
-		size, err := codec.EstimateValueSize(sc, r[id])
-		if err != nil {
-			continue
+	if sessVars.TxnCtx != nil {
```
Comment (siddontang) [inline table/tables/tables.go]: PTAL @coocood

**Comment by bobotu** [comment]
LGTM

**Comment by kennytm** [comment]
In the 33 GB TPC-C "CUSTOMER" CSV table test, the time needed to Encode() one 256 MiB chunk is reduced from ~28.5s to ~24.5s (86%), matching the expectation (Encode() doesn't just call AddRecord(), it also calls Convert(), so the ratio is higher than 73%).

However, the entire CSV → Importer step is only slightly improved from 2m37s to 2m34s (98%). This is because Encode() speed has caught up with the KV → Importer transfer speed, and is no longer the bottleneck.

File: table/tables/tables.go (hunk 1 above)
Comment (coocood) [inline table/tables/tables.go]: It would be simpler and has less diff if written like
```go
if sessVars.TxnCtx == nil {
    return recordID, nil
}
```

**Comment by kennytm** [comment]
@AilinKid something's wrong with the unit test.

```
[2020-03-02T07:09:15.193Z] FAIL: binloginfo_test.go:380: testBinlogSuite.TestBinlogForSequence
[2020-03-02T07:09:15.193Z] 
[2020-03-02T07:09:15.193Z] binloginfo_test.go:436:
[2020-03-02T07:09:15.193Z]     c.Assert(mustGetDDLBinlog(s, "select setval(`test2`.`seq2`, 6)", c), IsTrue)
[2020-03-02T07:09:15.193Z] ... obtained bool = false
```

**Comment by AilinKid** [comment]
> @AilinKid something's wrong with the unit test.
> 
> ```
> [2020-03-02T07:09:15.193Z] FAIL: binloginfo_test.go:380: testBinlogSuite.TestBinlogForSequence
> [2020-03-02T07:09:15.193Z] 
> [2020-03-02T07:09:15.193Z] binloginfo_test.go:436:
> [2020-03-02T07:09:15.193Z]     c.Assert(mustGetDDLBinlog(s, "select setval(`test2`.`seq2`, 6)", c), IsTrue)
> [2020-03-02T07:09:15.193Z] ... obtained bool = false
> ```
OK, I will check it.

**Comment by kennytm** [comment]
```
[2020-03-03T02:29:15.145Z] FAIL: builtin_time_test.go:764: testEvaluatorSuite.TestNowAndUTCTimestamp
[2020-03-03T02:29:15.145Z] 
[2020-03-03T02:29:15.145Z] builtin_time_test.go:798:
[2020-03-03T02:29:15.145Z]     c.Assert(ts.Sub(gotime(t, ts.Location())), LessEqual, 3*time.Millisecond)
[2020-03-03T02:29:15.145Z] ... compare_one time.Duration = 7432786 ("7.432786ms")
[2020-03-03T02:29:15.145Z] ... compare_two time.Duration = 3000000 ("3ms")
[2020-03-03T02:29:15.145Z] 
```

/merge

**Pull request merged**